pinot-core/src/main/java/org/apache/pinot/core/upsert/UpsertProcessorUtil.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.upsert;
+
+import java.util.Map;
+import org.apache.pinot.core.realtime.impl.ThreadSafeMutableRoaringBitmap;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public final class UpsertProcessorUtil {
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpsertProcessorUtil.class);
+
+  private UpsertProcessorUtil() {
+  }
+
+  public static void handleUpsert(PrimaryKey primaryKey, long timestamp, String segmentName, int docId, int partitionId,
```

Comment:
Suggest moving this logic into the `PartitionUpsertMetadataManager` so that all metadata management methods are at the same place

Reply:
hmm, I thought about that. But this logic does not fully belong to the `PartitionUpsertMetadataManager `, but only the part in the `else if` branch.

In fact, it's a common logic used in two places, and that's why I put it in a util class.

Reply:
I checked both places, and IMO the `handleUpsert` (update metadata) should definitely be part of the metadata manager.
The benefit of putting this into the manager class is that we can easily add more concurrency control for the class if needed (I'm pretty sure there are cases not handled)

Reply:
I don't quite buy the concurrency control argument. Static methods are executed concurrently, and if the concurrent access within this block of code is actually on `PartitionUpsertMetadataManager ` access. So from this front, the current way has better concurrency control IMO.

Code organization for better clarity is a different perspective, and we can discuss it.